### ISSUE TITLE
Work Logging causes Hang in NCCL Async Error Handling

### ISSUE THREAD

**Issue by osalpekar** [pytorch/pytorch#46278]
## 🐛 Bug

tl;dr - We must remove tensor-related logging from the `WorkNCCL::operator<<` function, otherwise printing the work objects tracked in `ProcessGroupNCCL::workMetaList_` will cause segfaults.

The Work objects we track in the workMetaList for the NCCL Async Error Handling mechanism don't contain output tensors (`WorkNCCL::outputs_`). As described in the `ProcessGroupNCCL::workEnqueue` function, destructing the output tensors calls into autograd_meta, which happens in the user thread, but our system destructs work objects in the workCleanupThread, so this could lead to a deadlock scenario. We avoid this problem by not tracking the tensors in the work objects that are added to the workMetaList (it's called work meta list because these work objects only track the metadata and not the actual tensors), so when the `WorkNCCL::operator<<` function tried to log tensor shapes for work objects from the watchdog thread, the async error handling mechanism hanged (in the desync test) or segfaulted (in the desync flow). 

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @xush6528 @osalpekar @jiayisuse @agolynski

**Comment by osalpekar** [comment]
Closed by #46265
